### PR TITLE
feat: Add minimal PySpark support

### LINKED ISSUE #333: [Enh]: Add Support For PySpark
### We would like to learn about your use case. For example, if this feature is needed  to adopt Narwhals in an open source project, could you please enter the link to it below?

_No response_

### Please describe the purpose of the new feature or describe the problem to solve.

PySpark is one of the most used dataframe processing frameworks in the big data space. There are whole companies built around it and it is commonplace in the Data Engineering realm.

One common pain point is that data scientists usually works with Pandas (or more recently Polars) and when integrating their code in big ETL processes, that code is usually converted to PySpark for efficiency and scalability.

I believe that is precisely the problem Narwhals tries to solve and it would be a great addition to the data ecosystem to support PySpark.

### Suggest a solution if possible.

PySpark has two distinct APIs:

- PySpark SQL - Identical API to Scala version of Spark
- PySpark Pandas - Subset of Pandas API (Previously known as Koalas)

Given that PySpark Pandas has an API based on Pandas, I believe it should be relatively straightforward to re-use the code already written for the Pandas backend.

There is a PySpark SQL to PySpark Pandas conversion so in theory it should be possible to also add ad hoc support for PySpark SQL Dataframes and check the overhead. If it is too big it can be considered to also add a separate backend for that different API.

### If you have tried alternatives, please describe them below.

_No response_

### Additional information that may help us understand your needs.

I do have experience with the PySpark API and would like to contribute, I read the "How it works" section but would like some concrete direction on how to get started and if this is of interest to the maintainers

### PR DESCRIPTION
**Pull request by EdAbati**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Closes #333

## Checklist

- [x] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

## If you have comments or can explain your changes, please do so below.

As mentioned in the latest call, I've started working on the support for PySpark. 

The goal of this PR would be to have a minimal initial implementation as a starting point. As we did for Dask, we can implement single methods in following PRs!

~⚠️ this is not ready for review, a lot of test are failing, the code is ugly. :) Just making the PR for visibility and to have a place to comment/ask questions on specific points~

### CODE REVIEW

**Comment by EdAbati** [comment]
This PR diff is getting big because of all the `xfail` in tests. 😕
@MarcoGorelli @FBruzzesi do you have a better idea on how to make it more "reviewable"?

**Comment by FBruzzesi** [comment]
> This PR diff is getting big because of all the `xfail` in tests. 😕 @MarcoGorelli @FBruzzesi do you have a better idea on how to make it more "reviewable"? or do you think it is fine?

For dask we started with it's own test file, so that we didn't have to modify every other file.
Once we had a few methods implemented we shifted the constructor into the conftest list of constructors and added the xfails.

Would that be a good strategy again?

**Comment by EdAbati** [comment]
This is finally ready for review 🥵

I have to fix the tests on windows (I think I may need to set up Java or something similar) and the test with old version of pandas (that is not compatible with PySpark)
But the rest should be ready!

I tried my best to keep it as small as possible while trying to implement the main functionality. Let me know what you think

File: narwhals/_pyspark/expr.py
Comment (EdAbati) [inline narwhals/_pyspark/expr.py:211]: Not sure this is ideal. Unfortunately the `stddev` in pyspark SQL does not support `ddof`
From `3.4` they added one function in the pandas namespace that supports that.

Based on which version one has, `std` may return a different result :( Any ideas?

**Comment by EdAbati** [comment]
@MarcoGorelli @FBruzzesi which criteria we used decide the minimal supported versions? Popularity? Time of release?

**Comment by MarcoGorelli** [comment]
i'd suggest, the lowest one that's not too difficult to support 😄 I think it'd be OK to set it quite high here, we can always work on lowering it later if there's demand

**Comment by EdAbati** [comment]
Because of the `pyspark` current dependencies, trying to make all the tests pass is a bit tricky.

I decided to make `3.3.0` as a minimum dependencies, the most recent version is `3.5.2`.
None of these is _technically compatible_* with Python 3.12. Is there a way to exclude it from tests without affecting coverage?

\* _technically compatible_: Ubuntu tests on 3.12 seem fine, Windows complains. Nevertheless Python 3.12 will be officially supported from 4.0.0: https://issues.apache.org/jira/browse/SPARK-44120

File: narwhals/_pyspark/expr.py
Comment (FBruzzesi) [inline narwhals/_pyspark/expr.py:136]: Would other require to be parsed into column (e.g. if it is an expr)? Same for all other dunder methods

**Review by FBruzzesi — COMMENTED** [review]
Hey @EdAbati this is an awesome effort! And the CI failing is just due to coverage and no failing test is even more incredible! 

I am not working with pyspark for some time, I tried to look at it with some critical eyes! I hope it helps

File: narwhals/_pyspark/namespace.py
Comment (EdAbati) [reply narwhals/_pyspark/namespace.py:76]: not the docs but FYI https://github.com/apache/spark/blob/master/python/pyspark/sql/column.py#L284-L296 ✌️

File: narwhals/_pyspark/dataframe.py
Comment (EdAbati) [reply narwhals/_pyspark/dataframe.py:98]: thanks! I have missed that. I had 3.2.0 as min dependency when I started on this :D

File: narwhals/_expression_parsing.py
Comment (EdAbati) [reply narwhals/_expression_parsing.py:39]: good point! 👌

File: narwhals/_spark/dataframe.py
Comment (MarcoGorelli) [inline narwhals/_spark/dataframe.py:42]: ooooh 😍 thinking ahead so we can get duckdb in here?

File: narwhals/_spark/dataframe.py
Comment (MarcoGorelli) [inline narwhals/_spark/dataframe.py:74]: should we leave `collect` out for now?

File: tests/spark_test.py
Comment (MarcoGorelli) [inline tests/spark_test.py:37]: does it work to add this to the `constructors` fixture and xfail whatever isn't supported yet? then we can see how much further we have to go

**Review by MarcoGorelli — COMMENTED** [review]
awesome work here @EdAbati

File: narwhals/_spark/dataframe.py
Comment (EdAbati) [reply narwhals/_spark/dataframe.py:74]: I think it is needed for collecting the results in the tests, right?  😕

File: tests/spark_test.py
Comment (MarcoGorelli) [reply tests/spark_test.py:38]: ah I see, thanks

sure, we could switch over as a follow-up?

File: narwhals/_pyspark/dataframe.py
Comment (EdAbati) [reply narwhals/_pyspark/dataframe.py]: `pandas` (as well as `pyarrow` and `numpy`) is still a dependency of PySpark https://github.com/apache/spark/blob/master/python/packaging/classic/setup.py#L350-L359

`toPandas()` will not fail, but I agree that it would be nice to give the user a choice.

To me, this sounds like a followup what do you think?

File: narwhals/_spark/dataframe.py
Comment (MarcoGorelli) [reply narwhals/_spark/dataframe.py]: ah right!

ok for now, if we don't publicize pyspark support we can make changes later anyway if necessary

File: narwhals/_spark/dataframe.py
Comment (EdAbati) [reply narwhals/_spark/dataframe.py]: I think so :)

File: narwhals/_spark/dataframe.py
Comment (EdAbati) [reply narwhals/_spark/dataframe.py]: Is there something in the current implementation that doesn't look ideal? (a part from this https://github.com/narwhals-dev/narwhals/pull/908#discussion_r1807362860)

**Comment by MarcoGorelli** [comment]
looks like there's some mysterious error in CI

```
ERROR tests/spark_test.py::test_std[_pyspark_constructor_with_session] - pytest.PytestUnraisableExceptionWarning: Exception ignored in: <function __install_postfork_hook.<locals>.before_hook at 0x7f32678c40d0>
```

If we can get CI green, I think it'd be OK to merge this so we can then incrementally improve and avoid the PR going stale? Well done sticking with this one 💪

**Comment by EdAbati** [comment]
I think it is because of a new warning introduced yesterday in `1.14` [source](https://github.com/pola-rs/polars/releases/tag/py-1.14.0). I need to check if we import polars in any code triggered by the spark workers (or similar)

File: narwhals/_pyspark/expr.py
Comment (EdAbati) [reply narwhals/_pyspark/expr.py]: thanks :)

File: narwhals/_pyspark/dataframe.py
Comment (EdAbati) [reply narwhals/_pyspark/dataframe.py]: can you make an example of what you mean?

are you referring to something like `df.select(nw.max("a"))` ?

**Comment by EdAbati** [comment]
Tests are green (a part from an unrelated one)

Please let me know if you or @FBruzzesi have any other questions or ideas how to improve this :)

File: narwhals/_spark_like/utils.py
Comment (EdAbati) [reply narwhals/_spark_like/utils.py:39]: good catch! I don't know what I was thinking tbh :D

File: narwhals/translate.py
Comment (MarcoGorelli) [inline narwhals/translate.py:728]:
```suggestion
            level="lazy",
```

**Comment by MarcoGorelli** [comment]
amazing, thanks for updating! happy to merge once it's green as 🥦

File: narwhals/_spark_like/utils.py
Comment (EdAbati) [reply narwhals/_spark_like/utils.py:57]: yes :)

File: narwhals/_spark_like/utils.py
Comment (EdAbati) [reply narwhals/_spark_like/utils.py:52]: great point!

I _think_ we cannot get it from the type itself but I need some more time to do some research.

If found things pyspark does when converting to pandas that could be useful to us:
- https://github.com/apache/spark/blob/master/python/pyspark/sql/pandas/types.py#L716
- https://github.com/apache/spark/blob/master/python/pyspark/sql/pandas/types.py#L607

Still have to spend some time to try to understand, but with a brief look it seems the time_unit is set to `ns` and time_zone is set to the local timezone.

For arrow it seems to be different though https://github.com/apache/spark/blob/master/python/pyspark/sql/pandas/types.py#L66

Comment (MarcoGorelli) [reply narwhals/_spark_like/utils.py:52]: sure, happy to defer this

**Comment by EdAbati** [comment]
Thank you all for the feedback and apologies this is taking longer but I am a bit busy.

The CI is _almost_ green 😅 I need to add some if-else because `pyspark` is not available in `python>3.11`. Hopefully it will be fixed tonight

regarding the issue with `df.with_columns(d=nw.col('a').mean()).collect().to_native()`
I may have an idea on how to do it. I try to squeeze this is in too (and add a test). If this takes longer, could it be a follow-up?

**Comment by MarcoGorelli** [comment]
> If this takes longer, could it be a follow-up?

yup definitely!

**Comment by EdAbati** [comment]
Ok CI is 🌲🌳

Regarding that issue, I am still working on a fix 🥲

**Review by MarcoGorelli — APPROVED** [review]
thanks @EdAbati for this massive effort

cool, let's ship it before merge conflicts creep it, excited about building this up!

**Comment by EdAbati** [comment]
Exciting 🎉🎉🎉 thanks!

**Pull request merged**
